Makefile

```diff
@@ -164,6 +164,51 @@ OCAML_NATIVE_PROGRAMS = ocamlnat tools/lintapidiff.opt
 $(foreach PROGRAM, $(OCAML_NATIVE_PROGRAMS),\
   $(eval $(call OCAML_NATIVE_PROGRAM,$(PROGRAM))))
 
+# OCaml libraries that are compiled in both bytecode and native code
+
+# List of compilerlibs
+
+COMPILERLIBS = $(addprefix compilerlibs/, \
+  ocamlbytecomp \
+  ocamlcommon \
+  ocamlmiddleend \
+  ocamloptcomp \
+  ocamltoplevel)
+
+# Since the compiler libraries are necessarily compiled with boot/ocamlc,
+# make sure they *always are*, even when rebuilding a program compiled
+# with ./ocamlc (e.g. ocamltex)
+
+$(COMPILERLIBS:=.cma): \
+  CAMLC = $(BOOT_OCAMLC) $(BOOT_STDLIBFLAGS) -use-prims runtime/primitives
+
+# FIXME: how about making another target depend on $(ALL_CONFIG_CMO)?
+compilerlibs/ocamlcommon.cma: $(ALL_CONFIG_CMO)
+
+OCAML_LIBRARIES = $(COMPILERLIBS)
+
+$(foreach LIBRARY, $(OCAML_LIBRARIES),\
+  $(eval $(call OCAML_LIBRARY,$(LIBRARY))))
+
+# Let the native libraries depend on the native compiler
+
+$(ocamlcommon_CMX_FILES) $(ocamlbytecomp_CMX_FILES) \
+$(ocamloptcomp_CMX_FILES) $(ocamltoplevel_CMX_FILES): ocamlopt$(EXE)
```

Comment:
I think it's probably an error that the middle end library wasn't in this list (it will be in practice - I'm sure each module in the middle-end in someone depends on ocamlcommon so were rebuilt regardeless), but it would simplify the rule I think if it were simply that all the .cmx files of all the compilerlibs depend on ocamlopt?